src/html/mod.rs

```diff
@@ -92,6 +92,23 @@ impl Html {
         }
     }
 
+    /// Fast select open edges that match selector into vector of elements
```

Comment:
I think an example that demonstrate the difference between the returned nodes for `select` and `matches` would be helpful. (If there is a difference that is. Otherwise, I am not sure why the separate methods are kept instead of just replacing the implementation of `Select`.)

Reply:
I agree, I think the ladder seems like a good option since the API is used the same way for the most part.

Reply:
> for the most part.

But is it the same, or only mostly the same?

Reply:
The same, and a lot faster. A lot less work needed for selecting elements.

Reply:
So then it should just replace the current `Select` implementation instead of being a separate method. (Return a named type is still preferable though as a courtesy to downstream users. But it should be possible to just embed `ego_tree::iter::Traverse` inside `Select` instead of `ego_tree::iter::Nodes`.)

Out of curiosity, could you explain why this is much faster than the nodes iterator while yielding exactly the same results?

Reply:
@adamreichold I think it has to do with side-effects. I did not look into the code that deeply but, was testing rigorously across multiple data sets and saw there was a bottleneck just on the the select iteration method internally when I broke it apart. Consistently across the board was over 60x - I think it could do with something on `drop` being performed.

Reply:
Are you measuring with LTO enabled? Or put differently, did you try marking the `fn next` implementation of `Select` as `#[inline]` to ensure that it get a similar treatment as the `impl Iterator` variant?

Reply:
I pushed up an example here https://github.com/spider-rs/spider/tree/perf/minitest you can run it with `cargo run --example example --release`.

```
HTML SIZE 4175
Time elapsed traverse() is: 62.25µs
Time elapsed in nodes() is: 666ns
- "https://rsseau.fr/"
```

Traversing is night and day faster if you convert the measurements to MS since the duration rounds it near the best counter measurement whole.

Reply:
```rust
        let html = Html::parse_document(self.html.as_str());
        tokio::task::yield_now().await;

        let start = std::time::Instant::now();

        println!("HTML Len {:?}", self.html.len());

        // extremely fast ego tree handling
        for node in html.tree.root().traverse() {
            match node {
                ego_tree::iter::Edge::Open(node_ref) => {}
                _ => (),
            }
        }
        let duration = start.elapsed();
        println!("Time elapsed traverse() is: {:?}", duration);

        let start = std::time::Instant::now();

        for node in html.tree.nodes() {}
        let duration = start.elapsed();

        println!("Time elapsed in nodes() is: {:?}", duration);
```

Reply:
I accidentally left the match statement in the traverse and still night and day faster.

Reply:
> Are you measuring with LTO enabled? Or put differently, did you try marking the `fn next` implementation of `Select` as `#[inline]` to ensure that it get a similar treatment as the `impl Iterator` variant?

testing now.

Reply:
Closing it is actually slower, @adamreichold thank you. Thank you for helping me get to this before heavy profiling that saved a lot of time!

Reply:
@cfvescovo @adamreichold I went and looked one last time and thought I knew something was funny. It was not the iteration it was the matching. I pushed up the changes to the test [here](https://github.com/spider-rs/spider/commit/6b1019bc1b33a6c51521f2c3904aafa711c53039).

Reply:
Could you elaborate a bit on that?

Reply:
> ~~~@cfvescovo @adamreichold I went and looked one last time and thought I knew something was funny. It was not the iteration it was the matching. I pushed up the changes to the test [here](https://github.com/spider-rs/spider/commit/6b1019bc1b33a6c51521f2c3904aafa711c53039).~~~ nvm again, pardon the noise.

Reply:
Don't worry, thank you for your time!

Reply:
@cfvescovo I realized what made me stumble on this change. Using `select` or walking with nodes prevents async from being used or streamed. Using the root tree allows await points during the link iteration.
 
![Screenshot 2023-03-01 at 4 47 33 PM](https://user-images.githubusercontent.com/8095978/222272305-b77bcd61-5623-40af-9832-7f487a04f120.png)

Reply:
logic for the `links_ego`.

![code for links egho](https://user-images.githubusercontent.com/8095978/222272765-34ed54c7-2178-4526-95bc-7ecf23d6d63d.png)

Reply:
I think the problem here is that we cannot provide a method like

```rust
fn into_select(self, selector: &Selector) -> IntoSelect;
```

on `Html` which used `ego_tree::iter::IntoIter` internally as we cannot match selectors against bare `Node`s as selectors need to "look around" in the tree. Hence this use case of "streaming extraction" would always have to fall back to manually selecting nodes based on their attributes as in your screenshot above. But of course, this is already supported by the current API due to `tree` being a public field.

As a general remark, did you try just iterating the links synchronously but `spawn`ing a separate task for each URL? Tasks are pretty cheap after all and this should give you more or at least the same concurrency and parallelism as the "streaming" version.

Reply:
@adamreichold you are right on a lot of points and the effectiveness on how cheap it is. Really the bottlenecks start to arise once there’s many concurrent tasks let’s say around 200 spawned with network IO switching between the tasks. The select statement performs really fast until it starts slowing down a bit and taking 1ms or so to iterate through which blocks.

For my situation I ended up having two APIs or entries for selecting links. 

When there’s no concurrent task for full sync and when there is to use the streaming handling.

Originally the solution I had was a task for the gathering of the html and including the select in a spawn full sync before any Atomic tendril changes. Being able to await after parsing the HTML fragment also helped which could have been a big factor too.

We might want to look into setting up some profiling since all of us seem keen to making this the best that it can be 😃.

@cfvescovo